packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1045,12 +1146,41 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
     return expirationTime;
   }
 
-  function scheduleWork(fiber: Fiber, expirationTime: ExpirationTime) {
-    return scheduleWorkImpl(fiber, expirationTime, false);
+  function retryOnPromiseResolution(
+    root: FiberRoot,
+    suspendedTime: ExpirationTime,
+  ) {
+    resumePendingWork(root, suspendedTime);
+    const retryTime = findNextExpirationTimeToWorkOn(root);
+
+    if (retryTime !== NoWork) {
+      requestRetry(root, retryTime);
+    }
+  }
+
+  function markTimeout(timeoutMs: number) {
+    if (timeoutMs >= 0 && timeoutMs < nextEarliestTimeoutMs) {
+      nextEarliestTimeoutMs = timeoutMs;
+    }
+  }
+
+  function waitForTimeout(root, ms, suspendedTime) {
+    setTimeout(() => {
+      retryOnPromiseResolution(root, suspendedTime);
+    }, ms);
```

Comment:
I bumped into a funny case where I accidentally clobbered the timeout queue with a bunch of `setTimeout`s that reschedule themselves, and the fallback never showed up for this reason. Was very tricky to debug.